analyzers/src/SonarAnalyzer.Common/Rules/PreferGuidEmptyBase.cs

```diff
@@ -0,0 +1,82 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class PreferGuidEmptyBase<TSyntaxKind> : SonarDiagnosticAnalyzer<TSyntaxKind>
+        where TSyntaxKind : struct
+    {
+        internal const string DiagnosticId = "S4581";
+
+        protected override string MessageFormat => "Use 'Guid.NewGuid()' or 'Guid.Empty' or add arguments to this GUID instantiation.";
+
+        protected PreferGuidEmptyBase() : base(DiagnosticId) { }
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                Language.GeneratedCodeRecognizer,
+                c =>
+                {
+                    if (NotAllowedGuidCtorArguments(c.Node, c.SemanticModel)
+                        && c.SemanticModel.GetSymbolInfo(c.Node).Symbol is IMethodSymbol methodSymbol
+                        && methodSymbol.ContainingType.Is(KnownType.System_Guid))
+                    {
+                        c.ReportIssue(Diagnostic.Create(Rule, c.Node.GetLocation()));
+                    }
+                },
+                Language.SyntaxKind.ObjectCreationExpressions);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                Language.GeneratedCodeRecognizer,
+                c =>
+                {
+                    if (NoParameter(c.Node)
+                        && c.SemanticModel.GetTypeInfo(c.Node).ConvertedType.Is(KnownType.System_Guid))
+                    {
+                        c.ReportIssue(Diagnostic.Create(Rule, c.Node.GetLocation()));
+                    }
+                },
+                Language.SyntaxKind.DefaultExpressions);
+        }
+
+        private bool NotAllowedGuidCtorArguments(SyntaxNode ctorNode, SemanticModel semanticModel)
+        {
+            var arguments = Language.Syntax.ArgumentExpressions(ctorNode).ToArray();
+            return arguments.Length == 0 || CreatesGuidEmpty(arguments, semanticModel);
+        }
+
+        private bool NoParameter(SyntaxNode defaultExpression) =>
+            !defaultExpression.Ancestors()
+                .Any(node => Language.Syntax.IsKind(node, Language.SyntaxKind.Parameter));
+
+        private static bool CreatesGuidEmpty(SyntaxNode[] arguments, SemanticModel semanticModel) =>
+            arguments.Length == 1
+            && semanticModel.GetConstantValue(arguments[0]) is { HasValue: true } optional
+            && optional.Value is string str
+            && Guid.TryParse(str, out var guid)
```

Comment:
Can be simplified, as `Optional.Value` will return just `null` if `HasValue` is `false` and `TryParse` can accept `null` as input.
```suggestion
            && Guid.TryParse(semanticModel.GetConstantValue(arguments[0]).Value as string, out var guid)
```